lib/Dialect/TTIR/IR/TTIROps.cpp

```diff
@@ -409,34 +414,140 @@ ::mlir::LogicalResult mlir::tt::ttir::MatmulOp::verify() {
   ::mlir::RankedTensorType inputAType = getA().getType();
   ::mlir::RankedTensorType inputBType = getB().getType();
   ::mlir::RankedTensorType outputType = getOutput().getType();
-  auto inputAShape = inputAType.getShape();
-  auto inputBShape = inputBType.getShape();
-  auto outputShape = outputType.getShape();
-  if (inputAShape.size() < 2) {
-    return emitOpError("Input A must be at least a 2D tensor");
+
+  llvm::ArrayRef<int64_t> outputShape = outputType.getShape();
+  llvm::SmallVector<int64_t> inputAShape(inputAType.getShape());
+  llvm::SmallVector<int64_t> inputBShape(inputBType.getShape());
+
+  // If input A is a vector (1D tensor), 1 is prepended to its dimension for the
+  // purpose of the matrix multiply. After the matrix multiply, the prepended
+  // dimension is removed.
+  if (inputAType.getRank() == 1) {
+    inputAShape.insert(inputAShape.begin(), 1);
   }
-  if (inputBShape.size() < 2) {
-    return emitOpError("Input B must be at least a 2D tensor");
+
+  // If input B is a vector (1D tensor), a 1 is appended to its dimension for
+  // the purpose of the matrix-vector product and removed after.
+  if (inputBType.getRank() == 1) {
+    inputBShape.push_back(1);
   }
-  if (inputAShape.size() != inputBShape.size()) {
-    return emitOpError("Input A and B must have the same rank");
+
+  // Verify that the input A is at least 1D tensor
+  if (inputAType.getRank() < 1) {
+    return emitOpError("Input A must be at least a 1D tensor");
   }
-  if (inputAShape.size() != outputShape.size()) {
-    return emitOpError("Input A and B must have the same rank as the output");
+
+  // Verify that the input B is at least 1D tensor
+  if (inputBType.getRank() < 1) {
+    return emitOpError("Input B must be at least a 1D tensor");
```

Comment:
Maybe move these checks to be first in line.

Reply:
Agree, thanks for suggestion!